astropy/cosmology/tests/test_funcs.py

```diff
@@ -0,0 +1,184 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+import sys
+import io
+
+import pytest
+import numpy as np
+
+from astropy.cosmology import core, funcs, realizations
+from astropy.cosmology.realizations import Planck13, Planck15, Planck18, WMAP5, WMAP7, WMAP9
+from astropy.units import allclose
+from astropy import units as u
+from astropy.utils.exceptions import AstropyUserWarning
+from astropy.utils.compat.optional_deps import HAS_SCIPY  # noqa
+
+
+@pytest.mark.skipif('not HAS_SCIPY')
+def test_z_at_value():
+    # These are tests of expected values, and hence have less precision
+    # than the roundtrip tests below (test_z_at_value_roundtrip);
+    # here we have to worry about the cosmological calculations
+    # giving slightly different values on different architectures,
+    # there we are checking internal consistency on the same architecture
+    # and so can be more demanding
+    z_at_value = funcs.z_at_value
+    cosmo = Planck13
+    assert allclose(z_at_value(cosmo.age, 2 * u.Gyr), 3.19812268, rtol=1e-6)
+    assert allclose(z_at_value(cosmo.lookback_time, 7 * u.Gyr), 0.795198375, rtol=1e-6)
+    assert allclose(z_at_value(cosmo.distmod, 46 * u.mag), 1.991389168, rtol=1e-6)
+    assert allclose(z_at_value(cosmo.luminosity_distance, 1e4 * u.Mpc), 1.36857907, rtol=1e-6)
+    assert allclose(z_at_value(cosmo.luminosity_distance, 26.037193804 * u.Gpc, ztol=1e-10),
+                    3, rtol=1e-9)
+    assert allclose(z_at_value(cosmo.angular_diameter_distance, 1500*u.Mpc, zmax=2),
+                    0.681277696, rtol=1e-6)
+    assert allclose(z_at_value(cosmo.angular_diameter_distance, 1500*u.Mpc, zmin=2.5),
+                    3.7914908, rtol=1e-6)
+
+    # test behavior when the solution is outside z limits (should
+    # raise a CosmologyError)
+    with pytest.raises(core.CosmologyError):
+        with pytest.warns(AstropyUserWarning, match=r'fval is not bracketed'):
+            z_at_value(cosmo.angular_diameter_distance, 1500*u.Mpc, zmax=0.5)
+
+    with pytest.raises(core.CosmologyError):
+        with pytest.warns(AstropyUserWarning, match=r'fval is not bracketed'):
+            z_at_value(cosmo.angular_diameter_distance, 1500*u.Mpc, zmin=4.)
+
+    # Test the "verbose" flag. Since this uses "print", ned to mod stdout
+    original_stdout = sys.stdout
```

Comment:
The standard pytest way to do this is using 
https://docs.pytest.org/en/stable/monkeypatch.html

If so, easiest to make this a new test function (generally, in astropy we have been putting a bit too many tests in our "unit tests"... I'm definitely guilty!)

Reply:
Very cool pytest fixture.